File: frontend/src/types/item.ts
export interface ItemProperty {
  property_name: number;
  property_name_resolved?: string;
  subtype?: number;
  subtype_resolved?: string;
  cost_table?: number;
  cost_value?: number;
  cost_value_resolved?: string;
  param1?: number;
  param1_value?: number;
  chance_appear?: number;
}

export interface ScriptVariable {
  name: string;
  var_type: number;  // 1=int, 2=float, 3=string
  value: number | string;
}

export interface ScriptVariableInput {
  name: string;
  var_type: number;  // 1=int, 2=float, 3=string
  value: number | string;
}

export interface ItemSummary {
  resref: string;
  name: string;
  base_item: number;
  cost: number;
  stack_size: number;
  identified: boolean;
}

export interface ItemDetail {
  resref: string;
  name: string;
  localized_name?: Record<string, unknown>;
  description?: string;
  localized_description?: Record<string, unknown>;
  tag: string;
  base_item: number;
  cost: number;
  additional_cost: number;
  stack_size: number;
  charges: number;
  cursed: boolean;
  identified: boolean;
  plot: boolean;
  stolen: boolean;
  palette_id?: number;
  comment?: string;
  properties: ItemProperty[];
  variables: ScriptVariable[];
  raw_data?: Record<string, unknown>;
}

export interface ItemListResponse {
  items: ItemSummary[];
  total: number;
  offset: number;
  limit: number;
}

export interface ItemSearchResponse {
  items: ItemSummary[];
  total: number;
}

// Types for item editing

export interface LocalizedStringUpdate {
  text?: string;
  string_ref?: number;
}

export interface ItemPropertyInput {
  property_name: number;
  subtype?: number;
  cost_table?: number;
  cost_value?: number;
  param1?: number;
  param1_value?: number;
  chance_appear?: number;
}

export interface ItemTemplateUpdate {
  // Basic fields
  name?: LocalizedStringUpdate;
  description?: LocalizedStringUpdate;
  desc_identified?: LocalizedStringUpdate;
  tag?: string;
  cost?: number;
  additional_cost?: number;
  stack_size?: number;
  charges?: number;

  // Flags
  identified?: boolean;
  plot?: boolean;
  cursed?: boolean;
  stolen?: boolean;

  // Properties (full replacement)
  properties?: ItemPropertyInput[];

  // Script variables (full replacement)
  variables?: ScriptVariableInput[];

  // Model parts (icon/appearance)
  model_part1?: number;
  model_part2?: number;
  model_part3?: number;

  // Advanced
  new_resref?: string;
  palette_category?: number;
}

export interface PaletteCategory {
  id: number;
  strref: number;
  name: string;
}

export interface PaletteCategoriesResponse {
  categories: PaletteCategory[];
  available: boolean;
}

// Types for item property editor dropdowns

export interface PropertyTypeOption {
  id: number;
  label: string;
  has_subtype: boolean;
  has_cost_value: boolean;
  cost_table: number;
}

export interface PropertySubtypeOption {
  id: number;
  label: string;
}

export interface PropertyCostValueOption {
  id: number;
  label: string;
}

export interface PropertyTypesResponse {
  properties: PropertyTypeOption[];
}

export interface PropertySubtypesResponse {
  subtypes: PropertySubtypeOption[];
}

export interface PropertyCostValuesResponse {
  cost_values: PropertyCostValueOption[];
}

File: backend/models/item.py
from pydantic import BaseModel
from typing import Optional, List, Any


class ItemProperty(BaseModel):
    """Represents an item property on an item."""
    property_name: int
    property_name_resolved: Optional[str] = None
    subtype: Optional[int] = None
    subtype_resolved: Optional[str] = None
    cost_table: Optional[int] = None
    cost_value: Optional[int] = None
    cost_value_resolved: Optional[str] = None
    param1: Optional[int] = None
    param1_value: Optional[int] = None
    chance_appear: Optional[int] = None


class ScriptVariable(BaseModel):
    """Represents a script variable from VarTable."""
    name: str
    var_type: int  # 1=int, 2=float, 3=string
    value: Any  # int, float, or str depending on var_type

    @property
    def type_name(self) -> str:
        """Return human-readable type name."""
        return {1: "int", 2: "float", 3: "string"}.get(self.var_type, "unknown")


class ScriptVariableInput(BaseModel):
    """Model for creating/updating a script variable."""
    name: str
    var_type: int  # 1=int, 2=float, 3=string
    value: Any  # int, float, or str depending on var_type


class ItemSummary(BaseModel):
    """Summary info for item list views."""
    resref: str
    name: str
    base_item: int
    cost: int
    stack_size: int
    identified: bool


class ItemDetail(BaseModel):
    """Full item details including properties."""
    resref: str
    name: str
    localized_name: Optional[dict] = None
    description: Optional[str] = None
    localized_description: Optional[dict] = None
    tag: str
    base_item: int
    cost: int
    additional_cost: int = 0
    stack_size: int
    charges: int = 0
    cursed: bool = False
    identified: bool = True
    plot: bool = False
    stolen: bool = False
    palette_id: Optional[int] = None
    comment: Optional[str] = None
    properties: List[ItemProperty] = []
    variables: List[ScriptVariable] = []
    raw_data: Optional[dict] = None


class ItemCreate(BaseModel):
    """Model for creating a new item."""
    resref: str
    name: str
    tag: Optional[str] = None
    base_item: int
    cost: int = 0
    stack_size: int = 1


class ItemUpdate(BaseModel):
    """Model for updating an item."""
    name: Optional[str] = None
    tag: Optional[str] = None
    cost: Optional[int] = None
    stack_size: Optional[int] = None
    identified: Optional[bool] = None
    charges: Optional[int] = None
    cursed: Optional[bool] = None
    plot: Optional[bool] = None


class LocalizedStringUpdate(BaseModel):
    """Model for updating localized string fields."""
    text: Optional[str] = None  # Direct text (language 0)
    string_ref: Optional[int] = None  # TLK reference


class ItemPropertyInput(BaseModel):
    """Model for creating/updating an item property."""
    property_name: int
    subtype: int = 0
    cost_table: int = 0
    cost_value: int = 0
    param1: int = 255
    param1_value: int = 0
    chance_appear: int = 100


class ItemTemplateUpdate(BaseModel):
    """Model for updating an item template."""
    # Basic fields
    name: Optional[LocalizedStringUpdate] = None
    description: Optional[LocalizedStringUpdate] = None
    desc_identified: Optional[LocalizedStringUpdate] = None
    tag: Optional[str] = None
    cost: Optional[int] = None
    additional_cost: Optional[int] = None
    stack_size: Optional[int] = None
    charges: Optional[int] = None

    # Flags
    identified: Optional[bool] = None
    plot: Optional[bool] = None
    cursed: Optional[bool] = None
    stolen: Optional[bool] = None

    # Properties (full replacement)
    properties: Optional[List[ItemPropertyInput]] = None

    # Script variables (full replacement)
    variables: Optional[List[ScriptVariableInput]] = None

    # Model parts (icon/appearance)
    model_part1: Optional[int] = None
    model_part2: Optional[int] = None
    model_part3: Optional[int] = None

    # Advanced (triggers rename/palette updates)
    new_resref: Optional[str] = None
    palette_category: Optional[int] = None
